src/main/scala/Util.scala

```diff
@@ -170,3 +171,9 @@ object DecoupledMux {
   def apply[T <: Data](sel: Bool, a: DecoupledIO[T], b: DecoupledIO[T]): DecoupledIO[T] =
     apply(sel, Seq(b, a))
 }
+
+class ClockedIO[T <: Data](gen: T) extends Bundle {
+  val clock = Output(Clock())
+  val bits = gen
+  override def cloneType: this.type = (new ClockedIO(DataMirror.internal.chiselTypeClone[T](gen))).asInstanceOf[this.type]
```

Comment:
This needs some explanation. Why can't you pass gen through? Related, why can't you rely on autoclonetype?

Reply:
> Why can't you pass gen through?

I'm not an expert on this, but I think its because using `gen` as both `val bits` and in `def cloneType` results in two instances of `ClockedIO` using the same instance of `gen`. This results in a chisel `RebindingException`.
Alternatively, 
```
val bits = DataMirror.internal.chiselTypeClone[T](gen)
override def cloneType: this.type = (new ClockedIO(gen)).asInstanceOf[this.type]
``` 
also works.

> why can't you rely on autoclonetype?

That relies on `gen` being accessible, but `(val gen: T)` results in a `AutoCloneTypeException`, with `constructor parameters (gen) have values that are hardware types, which is likely to cause subtle errors. Use chisel types instead:`.

Reply:
You can make gen be a `private val` parameter to fix the autoclonetype problem.

Do you expect gen will sometimes be bound to a hardware type? Do you care about the direction the bound thing has? If the answer is yes, and yes, then you'll need to use a fancier cloneType, otherwise good ole `cloneType` should work for this application.

I'm trying to understand why you need to call to DataMirror.internal (which shouldn't be necessary here). Did chisel complain about having both undirectioned and directioned fields in ClockedIO?

Reply:
I was playing around with this because i was curious. https://scastie.scala-lang.org/qYVPEuV7SfiXfAnh2ACu2g

Using a primitive type breaks everything if you don't wrap it in a direction, but that won't actually happen all that often if ever.

Reply:
The error is actually due to some interaction with a `cloneType` in the IOCell generators.

https://scastie.scala-lang.org/DChdc5wHQjO6TgGsYjMwvw

Reply:
ooo, that is weird. The assignment of gen to data is transitively creating a real binding for gen... If you remove the data field, or create a new node by cloning gen, it works fine. Pedantically, i think most scala-hygenic way to write this would be to have gen be a function parameter `() => T`, otherwise the direction preserving cloneType is required, which given rocketchip's reliance on chiselTypeClone is fine by me. Writing it this way will lead to the fewest surprises though:

```
class Clocked[T <: Data](private val gen: T) extends Bundle {
  val clock = Output(Clock())
  val data = DataMirror.internal.chiselTypeClone[T](gen)
}
```
https://scastie.scala-lang.org/pKendxz5SOSVoQf2jYxZKg